proxy/agent.go

```diff
@@ -0,0 +1,198 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"reflect"
+	"time"
+
+	"k8s.io/client-go/pkg/util/flowcontrol"
+
+	"github.com/golang/glog"
+)
+
+// Agent manages the restarts and the life cycle of a proxy binary.  Agent
+// keeps track of all running proxy processes and their configurations.  Hot
+// restarts are performed by launching a new Envoy process with a strictly
+// incremented restart epoch.  This matches Envoy semantics for restart epochs:
+// To successfully launch a new Envoy process that will replace the running
+// Envoy processes, the restart epoch of the new process must be exactly 1
+// greater than the highest restart epoch of the currently running Envoy
+// processes. The initial epoch is 0.
+//
+// Envoy hot restart documentation:
+// https://lyft.github.io/envoy/docs/intro/arch_overview/hot_restart.html
+type Agent interface {
+	// Apply sets the desired configuration for the proxy.  Agent compares the
+	// current active configuration to the desired state and initiates a restart
+	// if necessary. If the restart fails, the agent attempts to retry with an
+	// exponential back-off.
+	Apply(config interface{})
+
+	// Run starts the agent and awaits for a signal on the input channel
+	Run(stop <-chan struct{})
+}
+
+// NewAgent creates a new proxy instance agent for a "run" function is a
+// blocking call to start the proxy, "cleanup" function is called after "run"
+// finishes.  Both functions take the epoch as an argument. Retry is budgeted
+// and uses an exponential back-off from the initial delay.
+func NewAgent(run func(interface{}, int) error, cleanup func(int),
+	initialRetryBudget int, initialRetryDelay time.Duration) Agent {
+	return &agent{
+		run:                run,
+		cleanup:            cleanup,
+		epochs:             make(map[int]interface{}),
+		config:             make(chan interface{}),
+		status:             make(chan exitStatus),
+		initialRetryBudget: initialRetryBudget,
+		initialRetryDelay:  initialRetryDelay,
+	}
+}
+
+type agent struct {
+	// proxy start-up command
+	run func(interface{}, int) error
+
+	// proxy cleanup command
+	cleanup func(int)
+
+	// desired configuration state
+	desired interface{}
+
+	// active epochs and their configurations
+	epochs map[int]interface{}
+
+	// current configuration is the highest epoch configuration
+	current interface{}
+
+	// channel for posting desired configurations
+	config chan interface{}
+
+	// channel for proxy exit notifications
+	status chan exitStatus
+
+	// retry budget count
+	retryBudget int
```

Comment:
is this the current number of retries for the specific epoch? if so, I would rename this to retries and the initialRetryBudget to maxRetries. and retryDelay to initialRetryInterval. Although, i don't think we need that field to be configurable. Too many configuration parameters. We could just set a standard 5s initial retry delay.